Document/0x06c-Reverse-Engineering-and-Tampering.md

```diff
@@ -404,6 +404,60 @@ for armv7.
+
+![Obtain image base address using Ghidra](Images/Chapters/0x06c/debugging_ghidra_image_base_address.png "Obtain image base address using Ghidra")
+
+From our previous analysis in "[Manual (Reversed) Code Review](#manual-reversed-code-review)" section, in the UnCrackable Level 1 application, the value of the hidden string is stored in a label with *hidden* flag set. In the disassembly, the text value of this label is stored in register `X21`, `mov`d from `X0`, at offset `0x100004520`. This is our first address, _breakpoint offset_.
+
+![Breakpoint address using Ghidra](Images/Chapters/0x06c/debugging_ghidra_breakpoint.png "Breakpoint address using Ghidra")
+
+For the second address, we need to determine the ASLR shift offset for a given process. As mentioned before, iOS is a modern operating system with multiple techniques implemented to mitigate code execution attacks, one such technique being Address Space Randomization Layout (ASLR). On every new execution of an application, a random ASLR shift offset is generated, and various process' data structures are shifted by this offset. In order to set a breakpoint at the correct address we need to determine this shift offset each time. The ASLR offset can be determined by using `lldb` command `image list -o -f`. The output is shown in the screenshot below.
+
+![Process image list](Images/Chapters/0x06c/debugging_lldb_image_list.png "Process image list")
+
+In the output, the first column contains the sequence number of the image ([X]), the second column contains the randomly generated ASLR offset, while 3rd column contains the full path of the image and towards the end, content in the bracket shows the image base address after adding ASLR offset to the original image base address (`0x100000000` + `0x40000` = `0x100040000`). You will notice the image base address of `0x100000000` is same as in Ghidra. Now, to obtain the effective memory address for a code location we only need to add ASLR offset to the address identified in Ghidra. The effective address to set the breakpoint will be `0x100004520` + `0x40000` = `0x100044520`. The breakpoint can be set using command `b 0x100044520`.
+
+In the above output, you may also notice that many of the full paths of the images do not point to a path on the iOS device, instead it points to a location on the machine on which `lldb` is running. These are system libraries for which debug symbols are available on the machine (part of Xcode iOS development toolkit) to aid in application development and debugging. For these libraries we can use the function names directly to set breakpoints.
```

Comment:
Very interesting note here ;) just a little adaption from my side + I made it a "side note" as the following paragraph actually continues what you're telling in the previous to this one. 

```suggestion
> In the above output, you may also notice that many of the paths listed as images do not point to the file system on the iOS device. Instead, they point to a certain location on the machine on which lldb is running. These images are system libraries for which debug symbols are available on the machine to aid in application development and debugging (as part of the Xcode iOS SDK). Therefore, you may set breakpoints to these libraries directly by using function names.
```